super-ddf.c

```diff
@@ -4658,13 +4695,68 @@ static int ddf_set_array_state(struct active_array *a, int consistent)
 	if (old != ddf->virt->entries[inst].init_state)
 		ddf_set_updates_pending(ddf, NULL);
 
-	dprintf("ddf mark %d/%s (%d) %s %llu\n", inst,
-		guid_str(ddf->virt->entries[inst].guid), a->curr_state,
+	dprintf("ddf mark array%d guid:%s state:%s init_state:%s curr_state:%s:%d consistency:%s resync_start:%llu\n",
+		inst,
+		guid_str(ddf->virt->entries[inst].guid),
+		map_num(ddf_state, (ddf->virt->entries[inst].state & DDF_state_mask)),
+		map_num(ddf_init_state, (ddf->virt->entries[inst].init_state & DDF_initstate_mask)),
+		array_state_str(a->curr_state), a->curr_state,
 		consistent?"clean":"dirty",
 		a->info.resync_start);
 	return consistent;
 }
 
+/*
+ * ddf_update_state() Blockbridge
+ *
+ * This function retrieves the current array secondary state, for the
+ * first sub-array, and updates it if needed. In Blockbridge, only one
+ * subarray is ever possible, so these values are hard-coded here:
+ *
+ *	find_vdcr(ddf, 0, 0, ...)
+ *
+ * The second argument "0" is the first subarray. The third argument
+ * "0" is the first raid disk. We only need to get "any" vcl here. So
+ * this is always valid.
+ *
+ * Once the vcl is available, we check the state, compare it against
+ * the persistent state value. And schedule a metadata update if the
+ * current state is not reflected in the persistent state.
+ *
+ * When paired with array membership changes, calling this function
+ * ensures correct persistent "degraded" state, when necessary.
+ */
+static int
+ddf_update_state(struct supertype *st)
+{
+	struct ddf_super *ddf = st->sb;
+	unsigned int n_bvd;
+	struct vcl *vcl = NULL;
+	struct vd_config *vc = find_vdcr(ddf, 0, 0, &n_bvd, &vcl);
+	int state, old, upp;
+	if (!vc || !vcl)
+		return 1;
+
+	/* Now we need to check the state of the array and update
+	 * virtual_disk.entries[n].state.
+	 * It needs to be one of "optimal", "degraded", "failed".
+	 * I don't understand 'deleted' or 'missing'.
+	 */
+	state = get_svd_state(ddf, vcl);
+	old = (ddf->virt->entries[0].state & DDF_state_mask);
+	upp = (ddf->virt->entries[0].state & ~DDF_state_mask);
+
+	if (old != state) {
```

Comment:
yep, i see... `secondary_state` only returns shit in the first three bits, looks good.